File: parity/python_generator/src/c4a_parity_filters_ref/composite.py
# SPDX-License-Identifier: CECILL-2.1
"""
Composite sample filter — frozen NumPy reference.

The composite combines multiple sub-filter masks via either AND ("all") or
OR ("any") semantics, expressed in keep-mask space:

  * mode = "any" (default): a sample is EXCLUDED if ANY sub-filter excludes
    it.  Equivalent to ``keep = all(submask_k)``.
  * mode = "all":          a sample is EXCLUDED only if ALL sub-filters
    exclude it.  Equivalent to ``keep = any(submask_k)``.

The reference takes a list of pre-computed keep-masks (NumPy bool arrays of
identical shape) because the c4a engine doesn't replay the sub-filter
algorithms inside the composite — it holds references to the sub-filter
handles and aggregates their already-computed masks.

This matches ``nirs4all.operators.filters.CompositeFilter.get_mask``.
"""

from __future__ import annotations

from typing import Sequence

import numpy as np


def composite_mask(submasks: Sequence[np.ndarray],
                    *,
                    mode: str = "any") -> np.ndarray:
    """Aggregate sub-filter keep-masks via the given mode.

    Parameters
    ----------
    submasks : sequence of (n,) bool arrays
        Per-filter keep masks. Must have identical length.
        If empty, returns an all-True mask of inferred length 0.
    mode : "any" or "all"
        Mode semantics (see module docstring).

    Returns
    -------
    mask : (n,) bool array
    """
    if mode not in ("any", "all"):
        raise ValueError(f"mode must be 'any' or 'all', got '{mode}'")
    if not submasks:
        # Empty composite keeps all samples by convention.
        return np.ones(0, dtype=bool)
    arr = np.stack([np.asarray(m, dtype=bool) for m in submasks], axis=0)
    if mode == "any":
        return np.asarray(np.all(arr, axis=0))
    return np.asarray(np.any(arr, axis=0))

File: cpp/src/core/filters/composite.c
/* SPDX-License-Identifier: CECILL-2.1 */
/*
 * Composite filter reference implementation.
 *
 * Aggregation only — no algorithm replay.  See header for the contract.
 */

#include "composite.h"

#include <stddef.h>

n4m_status_t n4m_filter_composite_combine(
    int mode,
    const uint8_t* const* submasks,
    int n_sub,
    int64_t rows,
    uint8_t* mask_out,
    int64_t* out_n_samples,
    int64_t* out_n_kept,
    int64_t* out_n_excluded) {
    if (mask_out == NULL) {
        return N4M_ERR_NULL_POINTER;
    }
    if (rows < 0) {
        return N4M_ERR_INVALID_ARGUMENT;
    }
    if (mode != N4M_FILTER_COMPOSITE_ANY && mode != N4M_FILTER_COMPOSITE_ALL) {
        return N4M_ERR_INVALID_ARGUMENT;
    }
    if (n_sub < 0) {
        return N4M_ERR_INVALID_ARGUMENT;
    }
    if (n_sub > 0) {
        if (submasks == NULL) {
            return N4M_ERR_NULL_POINTER;
        }
        for (int k = 0; k < n_sub; ++k) {
            if (submasks[k] == NULL) {
                return N4M_ERR_NULL_POINTER;
            }
        }
    }

    int64_t kept = 0;

    if (n_sub == 0) {
        /* Empty composite keeps everything (matches the Python reference). */
        for (int64_t i = 0; i < rows; ++i) {
            mask_out[i] = 1;
        }
        kept = rows;
    } else if (mode == N4M_FILTER_COMPOSITE_ANY) {
        /* ANY = exclude if ANY subfilter excludes = keep iff ALL keep. */
        for (int64_t i = 0; i < rows; ++i) {
            uint8_t keep = 1;
            for (int k = 0; k < n_sub; ++k) {
                if (!submasks[k][i]) { keep = 0; break; }
            }
            mask_out[i] = keep;
            kept += keep;
        }
    } else {
        /* ALL = exclude only if ALL subfilters exclude = keep iff ANY keeps. */
        for (int64_t i = 0; i < rows; ++i) {
            uint8_t keep = 0;
            for (int k = 0; k < n_sub; ++k) {
                if (submasks[k][i]) { keep = 1; break; }
            }
            mask_out[i] = keep;
            kept += keep;
        }
    }
    if (out_n_samples)  *out_n_samples  = rows;
    if (out_n_kept)     *out_n_kept     = kept;
    if (out_n_excluded) *out_n_excluded = rows - kept;
    return N4M_OK;
}
